### ISSUE TITLE
Initializing New Editable Root List 

### ISSUE THREAD

**Issue by CrossSlide** [MarimerLLC/cslaforum#466]
WPF, Encapsulated Invocation, Editable Root List

When I Create a new List I want a specific set of child objects to be loaded. It would be just like a Fetch operation (parent passing DataReader to children) except the children would be marked as new. Is there a way to do this? 

Thanks

**Comment by jonnybee** [comment]
Yes, you can implement this in DataPortal_Create method and maybe pass a list of DTO/Poco objects for initialization of child objects to the child list Create method.

**Comment by CrossSlide** [comment]
Thanks @jonnybee. I'm not sure what you mean by passing objects to initialize. I want to retrieve the initialization data and create the children at the same time.

Below is the pattern I'd like to use but it's not working. Is there a flaw in my thinking on CSLA? It's just like the Fetch process.


`//In MyList
	public static MyList NewMyList(int Id)
        {
            return DataPortal.Create<MyList>(Id);
        }

	private void DataPortal_Create(int Id)
		{
			using (var dalManager = DalFactory.GetManager())
			{
				var dal = dalManager.GetProvider<IMyListDal>();
				using (var dr = dal.Create(Id))
				{
					while (dr.Read())
					{
						Add(DataPortal.CreateChild<MyChild>(dr));
					}
				}
			}
		}

//In IMyListDal		
	public interface IMyListDal
		{
			IDataReader Create(int Id);
		}
		

//In MyListDal
	public IDataReader Create(int Id)
        {
			//MySQLCode, get data with Id...
			return cm.ExecuteReader();
        }
		
//In MyChild
    private void Child_Create(System.Data.IDataReader data)
        {
            using (BypassPropertyChecks)
            {
                MyField = data.GetInt32(data.GetOrdinal("MyField"));
                MyField2 = data.GetInt32(data.GetOrdinal("MyField2"));
            }
        }
		`

**Comment by rockfordlhotka** [comment]
If I understand correctly you want to "fetch" a list of new objects. That really means you want to "create" a set of new objects that come pre-initialized with data.

This is the whole point of the DP_Create methods and functionality. So @jonnybee is right - you shouldn't "fetch" the list, you should "create" the list - and the criteria you pass to the `DataPortal.CreateAsync` method should be what you need in `DataPortal_Create` to initialize your list of new objects.

**Comment by CrossSlide** [comment]
Thanks @rockfordlhotka .

So if I understand you, then my DataPortal_Create (code above) can’t have the dalManager.GetProvider… and retrieve the initialization data I need. I would have to get the initialization data first then pass it to the DataPortal_Create. Correct?

**Comment by GillesBer** [comment]
@CrossSlide  I'm personally retrieving default values in the DataPortal_Create.
There is nothing that prevent to move your DataPortal_Fetch into your DataPortal_Create imho.

There are two points to take into consideration:
- if you need this code to run on the server/query the DB, you should not decorate your DataPortal_Create with the [RunLocal] attribute.
- having a round-trip to the server will incur a performance penalty (the same than the DataPortal_Fetch), which can be a problem if you perform lots of creation in your code.
Gilles

**Comment by rockfordlhotka** [comment]
Right. The only difference between DP_Fetch and DP_Create is how the metastate is set by the data portal.

DP_Create IsNew==true, IsDirty==true
DP_Fetch IsNew==false, IsDirty==false

Otherwise they work the same.

**Comment by CrossSlide** [comment]
@jonnybee @rockfordlhotka @GillesBer  Thanks guys I've got it working!
